src/main/java/jenkins/scm/api/SCMFileSystem.java

```diff
@@ -388,6 +441,67 @@ public static boolean supports(@NonNull SCMSource source) {
          */
         public abstract boolean supports(SCMSource source);
 
+        /**
+         * Checks if this {@link Builder} supports the supplied {@link SCMDescriptor}.
+         *
+         * @param descriptor the {@link SCMDescriptor}
+         * @return the return value of {@link #supportsDescriptor(SCMDescriptor)} if implemented, otherwise, it
+         * will return true if one of the following is true: the {@link SCM} for the descriptor is the enclosing
+         * class of this builder class, the builder and {@link SCM} are in the same package, or the builder is in
+         * a child package of the {@link SCM}.
+         * @since 2.3.0
+         */
+        public final boolean supports(SCMDescriptor<?> descriptor) {
+            if (Util.isOverridden(SCMFileSystem.Builder.class, getClass(), "supportsDescriptor", SCMDescriptor.class)) {
+                return supportsDescriptor(descriptor);
+            } else {
+                return descriptor.clazz.isAssignableFrom((getClass().getEnclosingClass())) ||
+                        getClass().getPackage().equals(descriptor.clazz.getPackage()) ||
+                        getClass().getPackage().getName().startsWith(descriptor.clazz.getPackage().getName());
+            }
+        }
+
+        /**
+         * Checks if this {@link Builder} supports the supplied {@link SCMDescriptor}.
+         *
+         * @param descriptor the {@link SCMDescriptor}
+         * @return {@code true} if and only if the supplied {@link SCMSourceDescriptor}'s {@link SCMDescriptor} class is
+         * supported by this {@link Builder}.
+         * @since 2.3.0
+         */
+        protected abstract boolean supportsDescriptor(SCMDescriptor descriptor);
```

Comment:
Won't this break any existing `Builder` implementations that hasn't yet adapted when upgrading?!

Reply:
If the logic is correct and the access is guarded, it should only cause compiler errors on updating the dependency... which is what we'd want, as if we could guess the answer we wouldn't need the method.

Of course the question for code review is has this been done correctly

Reply:
That does raise @jglick's question - what's the reasoning for this approach versus a single non-abstract method with the default heuristic that implementations can override?

Reply:
When I was doing my PhD my supervisor was notorious for redlining drafts of papers. People would routinely go through 50 or 70 revisions of their papers.

For my first paper, after 15 odd revisions I started to notice that every even numbered revision was basically the same and the odd numbered revisions were basically the same...

The redlines with corrections were oscillating back and forth... so revision 16 I didn’t take his corrections verbatim but instead used it as a hint to rewrite the sentence from scratch. Version 18 was the one we sent for publication and my subsequent papers were only 4-5 revisions.

You seem to be blindly following suggestions I make on my phone rather than stepping back and trying to figure out why i’m Prodding at these things (much like I was just taking my supervisor’s redlines verbatim)

I hope this story helps you

Reply:
...so I guess I took “Ideally what we would do is make this method final and make the SCMSource version non-abstract. Then we introduce protected abstract methods that the public final methods call (guarded by a reflective check of non-implemented)” too literally? And that doesn’t answer my question as to why you wanted this approach in the first place. So, no, the story is not helpful. =)

Reply:
So one of the lessons I learned in SCM-API is that there is great benefit in separation of the API from the SPI.

In SCM API the SPI (i.e. what implementers are supposed to implement) should be protected methods (to prevent consumers from calling them) whereas the API can be public methods on the SCM API API.

Given that we want to change/enhance behaviour for this method, that says that this is no longer a simple API, so we'd want to migrate more fully to the SPI vs API separation.

SPI evolution is different from API evolution. the SPI can break compile compatibility as long as runtime behaviour falls through. So in general that means ensuring that there is only one abstract method for anything rather than 5 overloads that people have to puzzle which to implement. If we need to add parameters, we move the `abstract` from one `protected` method to another. Then the `public final` method does the disambiguation and falls back through the potential implementations until it finds one that works or returns the "sensible" response.

Reply:
so...does that mean I should rollback to the final/abstract combo? My instinct is to go with a simple single overrideable method - but this is your turf, not mine, so I am not in a position to dictate the path forward, you are.